tensorflow-keras/src/main/java/org/tensorflow/keras/optimizers/OptimizerInterface.java

```diff
@@ -0,0 +1,61 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.keras.optimizers;
+
+import java.util.Map;
+import org.tensorflow.Graph;
+import org.tensorflow.op.Ops;
+
+/** The main Interface for Keras Optimizers */
+public interface OptimizerInterface {
+
+  /** The value for the name key in the Config object */
+  String NAME_KEY = "name";
+
+  /**
+   * Get a TensorFlow Graph from the Ops.
+   *
+   * @param tf the TensorFlow Ops
+   * @return the graph
+   * @throws java.lang.IllegalArgumentException if the TensorFlow Ops does not represent Graph mode
+   */
+  static Graph assertGraph(Ops tf) {
+    if(!tf.scope().env().isGraph()) {
```

Comment:
missing space after `if`

Reply:
OK, on the formatting,

If I refactor `OptimizerInterface` into` org.tensorflow.framework.optimizers.Optimizer,` then where should `assetGraph()` method go?  It is basically a utility that can be called from a `super()` ctor method, ensuring that `Ops tf` always represents a `Graph`, and then passes the `Graph` to the framework optimizers ctors.

Reply:
I'm fine with either or. 

I myself have some trouble figuring out how the `framework` and `keras` layer should split the job but at a high level, what I'm thinking is that most of the logic should occur in the framework and the core libraries while the Keras library should add a Pythonic Keras-like API as a facade to them. But I'm sure there will be exceptions. So in this case, if you don't think Keras users will need to access directly the graph of an optimizer and no such getter exists in the Python Keras lib, then maybe we should move it to the framework.

Reply:
Main issue is  consistency. The rest of the Keras apis are using `Ops` instead of `Graph` directly. Using `Ops` has an advantage, from my point of view, in that it insulates from a tight binding to `Graph`, when  one wants Eager support.

Reply:
Your comment also  brings up another consideration. If you take the premise that "_Keras library should add a Pythonic Keras-like API as a facade to them_", then why wouldn't `activations`, `loss`, `metrics`, `initializers`, etc. be primarily in frameworks rather than solely in Keras?

Reply:
...which is a totally legitimate question. It is not the first time we are questioning the roles of the framework and the Keras libraries, and the reason why we have both.

So initially the idea was what I explained, which is:
* core: exposes API of the native library in Java
* framework: high(mid)-level layer enriching this API with common ML utilities
* Keras: replicates the Keras API of Python on top of both

But now what we see is that we started to do some mid-level wrappers in the core directly (e.g. `sparseCrossEntropyWithLogits`) as we want to benefit from the `Ops` annotation processor exclusive to the core. Plus, the actual state of the TF API in Python is a little bit counter-intuitive as you cannot do everything just using the Keras API, as a result of the existence of the legacy API still present to this day. Since in Java we start everything from scratch, why not just doing a enhanced Keras API that can handle everything at the user perspective? 

As a Keras user myself, I'm always looking to use exclusively this API which I prefer to the old one and find it a bit incoherent and disappointing when I see that a few things are only possible to do with using vanilla TF. Why not keeping it simpler in Java and put everything in Keras then? It is questionable. 

It sounds to me that too that it would be easier for our collaborators if everything is one place instead of having to ask themselves how it should be split between these two similar-libraries...

Sorry if I'm not answering your question and probably bring more confusion but I think we need to set this up correctly to start with. Let's bring more people to this conversation: @Craigacp , @KartikChugh, @deansher

Reply:
To me, it seems worth writing down carefully chosen goals for each of the layers. Just to illustrate the sort of confusion I worry about, when @karllessard says "the Keras library should add a Pythonic Keras-like API as a facade to them", I ask myself these questions:
* Why didn't we want a Pythonic Keras-like API for the framework and core libraries in the first place?
* Given that we didn't, what's our goal in providing such an API?

Our existing  top-level `README.md` takes a shot at defining goals, but I don't think it goes far enough:

* `tensorflow-core`
  * All artifacts that build up the core language bindings of TensorFlow for Java. 
  * Those artifacts provide the minimal support required to use the TensorFlow runtime on a JVM.
    
* `tensorflow-framework`
  * High-level APIs built on top of the core libraries to simplify neural network training and inference 
    using TensorFlow.

I'm guessing what we need goes something like this:

* The core libraries should provide the most essential capabilities for someone who is strong in all three of Java, Tensorflow, and deep learning. It's hard to foresee why anyone besides this project would code entirely against this layer in the future, but it should be possible in principle, mostly just so the framework can be understood on its own.

* The framework should provide a comfortable deep learning framework for someone who is strong in both Java and deep learning, wants fine control throughout their modeling, and is happy to learn considerable Tensorflow along the way.

* The Keras layer should provide a comfortable deep learning API for someone who is fluent (maybe not super strong) in Java, is beginner to expert in deep learning, and wants to rely mainly on established best practice rather than controlling very detail. It should also be a reasonably comfortable transition for someone who knows the Python Keras API, although not at the expense of being clean, idiomatic Java.

Reply:
Just an FYI, the Keras Optimizers end up just being a thin facade over the framework Optimizers. The only difference may end up being the Keras defined parameter default values used in the Keras constructor that is passed to the framework constructor. However, I still think there is value in including classes in familiar locations. like `org.tensroflow.keras.optimizers`, even though they are a mere facade to the corresponding framework class..

Reply:
> Just an FYI, the Keras Optimizers end up just being a thin facade over the framework Optimizers.

Yes, and that is what they should be to fit in the current state of the project. 

I think @deansher did a pretty good summary of who's the audience for each library we distribute. The first customers of the core library I can think of is all the integrators of TensorFlow (DL4J, DJL, Tribuo...) that already have their own set of user libraries and API, so all they want/need is a thin layer to access the TensorFlow runtime from the JVM. I think we are actually doing good with this one. 

Now about the framework and Keras, do we really need both? If we do, then yes probably we want all initializers, metrics, etc. to be first added to the framework and then wrapped in Keras with a simplified interface, like @JimClarke5  is doing here with optimizers. But I'm wondering why a user needs to choose between an advanced API (framework) or a user-friendly one (Keras) instead of just having everything in a single library. And this might simplify/speedup our development as well.

I'm still just throwing ideas here, for the sake of brainstorming, but maybe there should not be a Keras library at all and that everything should end up in the framework, to which we would now give a "Keras" taste without being bound to strictly mirror the Python implementation?

Reply:
Actually my plan for Tribuo is that we'll allow users to define their graphs and models in TF, and then take over the presentation of training data batches, and wrap the trained model in Tribuo's interface. Depending on how the Keras interface turns out, we might be able to accept that, or we'll allow users to use the things in frameworks to define the model structure & training algorithm.

Reply:
@Craigacp when you say "my preference is to have both a low and high level framework", you clearly have in mind different audiences, missions, and/or characteristics of the two. Could you articulate those?

Reply:
And @JimClarke5 , based on your current experience, do you have any preference on this matter?

Reply:
IMHO, the beauty of Keras is in the simple, straight forward, `Model` and `Layers`. Most of the `Layers` have defaults for constructs like `Metrics`, `Optimizers`, `Activations`, etc. Also, they allow simple strings in their parameters that instruct the underlying layers  to construct elements, like `new Dense(24, "relu")`, so the way these elements are constructed can be hidden from a Keras user.   The fact that `Optimizers`, or other classes like `Metrics`, `Loss`, etc. are in `org.tensorflow.framework` versus `org.tensorflow.keras `does not seem that compelling to me. These elements can easily stand on their own outside of Keras. To a more sophisticated user, using `org.tensorflow.framework.metrics.Mean` to construct a specialized version of a `Metric`,  is not that different to including `org.tensorflow.keras.metrics.Mean`. The advantage of moving these elements to `framework`, is it would be clearer that they can be used in other contexts outside of Keras. The main disadvantage is that someone who is familiar with current Keras, would expect similar package names. However, this could be handled with the appropriate JavaDoc comments. So, I am leaning to add these elements to `framework`, similar to the way `optimizers` are currently postitioned. 

The next question would then be, where is the appropriate place for `Model` (and `Sequential`) and the numerous layers, like `Dense`, `Dropout`, `LSTM`, etc.? These too could be positioned under framework in packages like `org.tensorflow.framework.keras.model` and `org.tensorflow.framework.keras.layers` .

Reply:
By low level framework I mean the core ops, plus the optimizers that annotate the graph with the appropriate gradient optimizer, and potentially the dataset interface. That's what I'm likely to use for research, and also is probably the simplest to consume from Tribuo. Using the core ops directly without the convenience of the graph walking that the optimizers do means that the user will have to write all the graph walking code themselves which is probably too low level for most users. I think we should expose generic graph walking and modification operations at the framework level too to make it simpler for users who do want to do most things themselves.

The high level framework is for people who use Keras in TF Python, and want an API that guides them better. I think that we should have stronger typing information than exists in Python, as it's what would be expected from idiomatic Java and it helps IDEs & discoverability. For example, TF Keras in python has "activations" but these are bare functions in the activation package and specified by a String as an argument to a layer. In Java we should have a concrete interface that the activations extend from (even if they also expose the underlying function which directly applies the operation or writes it to the graph), and have the layer accept a strongly typed thing (I think this is the direction Jim is currently taking).

Reply:
Here's my shot at capturing this discussion. Perhaps we could reach consensus on this level of description and then I'll PR it into the README?

* `tensorflow-core`
  * All artifacts that build up the core language bindings of TensorFlow for Java
  * Intended audience: projects that provide their own APIs or frameworks on top of 
    TensorFlow and just want a thin layer to access the TensorFlow runtime from the JVM 
    
* `tensorflow-framework`
  * Complete but fairly primitive API for building and training neural networks with TensorFlow
  * Intended audience: expert neural network developers who prefer to make explicit, detailed decisions about their models and training algorithms
    
* `tensorflow-keras`
  * Partially covers the framework API to allow simpler definition of models and training algorithms
  * Intended to be familiar if you know the Python Keras API, but prioritizes clean, idiomatic Java over fidelity to Python
  * Provides defaults based on common best practices
  * Allows developers to selectively be more explicit by overriding defaults or dipping into the framework API
  * Intended audience: neural network developers across the spectrum from beginner to expert who prefer to rely mostly on best-practice defaults that they selectively fine-tune
  
* `ndarray`
  * Generic utility library for n-dimensional data I/O operations
  * Used by TensorFlow but does not depend on TensorFlow
  * Intended audience: any developer who needs a Java n-dimensional array implementation, whether or not they use it with TensorFlow

Reply:
> I think that we should have stronger typing information than exists in Python, as it's what would be expected from idiomatic Java and it helps IDEs & discoverability

I totally support that as well, we should avoid users to pass arbitrary string labels as parameters in the Java implementation. We could though have some sort of enums per components (optimizers, metrics, ...) which will restrict the possible choices without the need for the user to instantiate explicitly that component, to get closer to what the Python implementation offers. e.g. `model.compile(Optimizers.ADAM, Losses.BINARY_CROSSENTROPY)`

> Allows developers to selectively be more explicit by overriding defaults or dipping into the framework API

This is where I'm hesitating. If we allow users to access lower level functionalities from the Keras API, why do we need another API then? The current Python API is made up of two layers because it is historically the product of a merge between two different projects: the original TF API and the Keras project. I personally think it brings more confusion to the users that benefits and we don't need to follow this schema if we think we can do better in Java since we start from scratch.

I'm slowly leaning now to the idea of having a single API that supports both "beginner" and "advanced" modes, whether we call it Keras or not. At the same time, I don't want to slow @JimClarke5 down any further with this PR. Maybe let's continue that discussion with a broader audience? We can raise an new issue in GitHub, start another discussion on the mailing list or we could do this in the scope of a PR created by @deansher as he suggested (these descriptions sound good), what do you think?

Reply:
I think a PR to add that to the readme sounds good, and after that's merged we can discuss it in an issue.

My worry about forcing everything through something that's like Keras is that I train models that won't fit into a keras style model.fit call. One of the things I'm currently looking at is [USE](https://arxiv.org/abs/1803.11175), which is trained on multiple tasks, with each task using a different combination of dataset and loss function (e.g. train a question answering loss on one dataset, a textual similarity loss on another, and an MLM loss on a third, where apart from the top layer all layers are shared). However this isn't something that is well supported by any framework I've come across (unfortunately Google didn't release any training details or code on USE), and requires a lot of complexity that most users won't need. I can just about do it in bare TF or pytorch (in python, not tried using TF Java yet as we'd need TF text to work), but anything higher level requires fighting the framework at each step because I need control of the training loop.

Reply:
To move forward, may I suggest the following:
1) Add the 3 new Optimizers, `Adamax`, `Ftrl`, and `Nadam`,to `framework`.
2) Move the test cases to `framework`.
3) Hold off on the remaining classes in `org.tensorflow.keras.optimizers`, which are just a facade on `framework`, until we decide what to do with Keras. 

The next PR that I will work on will allow for a changing `LearningRate` and only impacts Optimizers in `framework`.
After that, I propose adding the other Keras elements, `Initializers`, `Activations`, etc., to `framework`.

Reply:
One other issue, is Keras defines the Optimizer `SGD` which is actually just `Momentum` in framework, should we just ignore this for the time being?

Reply:
> * Add the 3 new Optimizers, `Adamax`, `Ftrl`, and `Nadam`,to `framework`.
> 
> * Move the test cases to `framework`.
> 
> * Hold off on the remaining classes in `org.tensorflow.keras.optimizers`, which are just a facade on `framework`, until we decide what to do with Keras.

Sounds good @JimClarke5 , as in both approaches we are currently looking at, none of the optimizer backends will end up in the Keras module. Let's continue this discussion on #109 as it is becoming soon a priority that we all agree on this.

`SGD` can be added later, yes.